### PR TITLE
support save & restore of UB matrix reflections

### LINKED ISSUE #40: record wavelength as part of UB-defining reflections
The wavelength at which the r1/r2 were collected is not reported [in `pa()`]. It seems like the reflection does not hold the energy at which the angles were measured. Does the call to `compute_UB()` assume that the reflections were measured at the current energy?

_Originally posted by @gfabbris in https://github.com/BCDA-APS/apstools/issues/434#issuecomment-700923902_

### LINKED ISSUE #50: how to save/restore UB matrix?
Document the steps to save and to restore the UB matrix so that a session can be resumed.

### PR DESCRIPTION
**Pull request by prjemian**
fix #40 and fix #50

### CODE REVIEW

**Comment by prjemian** [comment]
@mrakitin @ambarb @gfabbris @jwkim-anl @strempfer -- Ready for review

**Review by gfabbris — APPROVED** [review]
LGTM

**Comment by prjemian** [comment]
@gfabbris : Thanks for the review!

**Comment by gfabbris** [comment]
I looked a bit more into (in the context of making fixQ energy scans) and it seems like that adding the diffractometer to the detectors will plot all the positioners in it. This is because the diffractometer hints will just get the `hints` field from each component (it gets it from `ophyd.Device` I think):

```python
@property
def hints(self):
    fields = []
    for _, component in self._get_components_of_kind(Kind.normal):
        c_hints = component.hints
        fields.extend(c_hints.get('fields', []))
    return {'fields': fields}
```

I wonder what is the best way around this issue. It looks like that modifying the diffractometer `hints` can fix it (see below), but maybe there is a better solution?

```python
@property
def hints(self):
    fields = []
    for _, component in self._get_components_of_kind(Kind.hinted):
        if (~Kind.normal & Kind.hinted) & component.kind:
            c_hints = component.hints
            fields.extend(c_hints.get('fields', []))
    return {'fields': fields}
```

**Comment by prjemian** [comment]
Might be the best we can do now but I'm uncomfortable with the entire kind+hints+visualization process.  Its documentation is minimal at best and very hard to locate.

Likely this *fix* will need to be revisited when the BestEffortCallback is revised or replaced.  A note here could help for that eventuality.

**Comment by ambarb** [comment]
This is something I wanted to visit, but this should be in Bluesky, right?  It has nothing to do with hklpy
I have list of things to test, but it seems to me that bec.peaks needs to rely on index and look up the whatever you want,

**Comment by prjemian** [comment]
@ambarb:  Agreed.  The code from @gfabbris () is not specific to hklpy.  We_can_ implement it here but it is better suited to the general discussion in *ophyd* but involves how *bluesky*'s BestEffortCallback handles this information.

**Comment by ambarb** [comment]
@gfabbris is correct. plotting is not sustainable or useful, but LiveTable is coupled to LivePlot is coupled to bec.  And not just for fixed Q E scan but for any diffractometer scan as it will be easier coding wise to put the entire diffractometer object in the detector list so you don't have to write a bunch of flow control on the data analysis side.

As for EfixQ, CSX's custom plan does not rely on having diffractometer motors in the detector list, but we did this for backwards compabiltiy if we want to use a different UB than the one at the time of collection.

**Comment by prjemian** [comment]
@mrakitin You weren't marked previously for review.  Would you like to review?

**Comment by mrakitin** [comment]
Hi @prjemian, I'll review it in the second part of the week. Mon-Wed are pretty busy already.

File: hkl/diffract.py [hunk 1]
```diff
@@ -34,7 +34,70 @@
...
+            self.constraint = constraint
+            self._index = 0
+
+        def __next__(self):
+            if self._index < len(self.constraint._fields):
+                c = getattr(self.constraint, self.constraint._fields[self._index])
+                self._index += 1
+                return c
+            else:
+                raise StopIteration
```
Comment (mrakitin) [inline hkl/diffract.py:91]: Class inside the class? Interesting, never encountered that. So it automatically becomes an attribute of the upper-level class?

File: hkl/diffract.py [hunk 2]
```diff
@@ -34,7 +34,70 @@
...
+
+    def __iter__(self):
+        "Iterate through the fields."
+        return self.ConstraintIterator(self)
+
+    def __repr__(self):
+        "Return a nicely formatted representation string."
+        content = "(" + ", ".join([f"{k}={getattr(self,k)}" for k in self._fields]) + ")"
```
Comment (mrakitin) [inline hkl/diffract.py:99]:
```suggestion
        content = "(" + ", ".join([f"{k}={getattr(self, k)}" for k in self._fields]) + ")"
```

Comment (mrakitin) [reply hkl/diffract.py:99]: Isn't it something that can use `self._repr_fmt` defined in `__init__`?

File: hkl/diffract.py [hunk 3]
```diff
@@ -53,7 +116,16 @@ class Diffractometer(PseudoPositioner):
         ~undo_last_constraints
         ~pa
         ~wh
+        ~_calc_energy_update_permitted
+        ~_constraints_dict
+        ~_constraints_for_databroker
         ~_energy_changed
+        ~_energy_changed
```
Comment (mrakitin) [inline hkl/diffract.py:123]: `~_energy_changed` appears twice. Is it intended?

File: hkl/diffract.py [hunk 4]
```diff
@@ -53,7 +116,16 @@ class Diffractometer(PseudoPositioner):
...
+        ~_energy_offset_changed
+        ~_energy_units_changed
+        ~_push_current_constraints
+        ~_set_constraints
+        ~_update_calc_energy
         ~_update_calc_energy
```
Comment (mrakitin) [inline hkl/diffract.py:129]: `~_update_calc_energy` appears twice. Is it intended?

File: hkl/tests/test_save_restore_UB.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+    for key in key_list:
+        key_name = f"fourc_{key}"
+        assert hasattr(xarray_data, key_name)
+        assert key_name in descriptors["data_vars"]
+    assert xarray_data.fourc_class_name == "Fourc"
+    assert xarray_data.fourc_geometry_name == "E4CV"
+    assert xarray_data.fourc_diffractometer_name == "fourc"
+    assert xarray_data.fourc_sample_name == "Si"
+    numpy.testing.assert_array_equal(xarray_data.fourc__pseudos, ["h k l".split()])
+    numpy.testing.assert_array_equal(xarray_data.fourc__reals, ["omega chi phi tth".split()])
```
Comment (mrakitin) [inline hkl/tests/test_save_restore_UB.py:114]: What is returned by `xarray_data.fourc__reals`? Is it a list of a list? The right-side argument is a double-list (`[['omega', 'chi', 'phi', 'tth']]`).

Also, I'm curious why `numpy.testing.assert_array_equal` is needed for compare strings?

File: hkl/tests/test_save_restore_UB.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    refls = fourc_orient["reflections_details"]
+    assert len(refls) == 2
+    assert "wavelength" in refls[0]
+    assert refls[0]["wavelength"] == 1.54
+
+
+def test_list_orientation_runs(fourc, kappa):
+    cat = databroker.temp().v2
+    RE = RunEngine({})
+    RE.subscribe(cat.v1.insert)
```
Comment (mrakitin) [inline hkl/tests/test_save_restore_UB.py:140]: `RE` and `cat` can be passed as fixtures to this test and the ones above, so the same code won't be replicated in many places.

File: hkl/tests/test_save_restore_UB.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+    # geometry mismatch, cannot restore
+    k4cv = Kappa("", name="k4cv")
+    assert len(k4cv.calc._samples) == 1
+    assert len(k4cv.calc.sample.reflections) == 0
+    with pytest.raises(ValueError) as exinfo:
+        hkl.util.restore_orientation(fourc_orient, k4cv)
+    args = exinfo.value.args
+    assert len(args) == 1
+    expected = "Geometries do not match: " "Orientation=E4CV, k4cv=K4CV, will not restore."
```
Comment (mrakitin) [inline hkl/tests/test_save_restore_UB.py:192]: Looks like `black` auto-formatting.
```suggestion
    expected = "Geometries do not match: Orientation=E4CV, k4cv=K4CV, will not restore."
```

File: hkl/util.py [hunk 8]
```diff
@@ -120,11 +147,262 @@ def get(attr):
...
+    Parameters
+    ----------
+    orientation : dict
+        Dictionary of orientation parameters (from
+        :func:`~hkl.util.run_orientation_info()`) recovered from run.
+    diffractometer : :class:`~hkl.diffract.Diffractometer()`
+        Diffractometer object.
+    """
+    _check_geometry(orientation, diffractometer)
+    # remeber this wavelength
```
Comment (mrakitin) [inline hkl/util.py:322]:
```suggestion
    # remember this wavelength
```

**Review by mrakitin — APPROVED** [review]
Nice addition! https://github.com/bluesky/hklpy/blob/80236f2603811b777286c2aab81de07bf94b9e93/examples/tst_UB_in_stream.ipynb was interesting reading.

**Comment by prjemian** [comment]
@mrakitin I'll make changes before the merge, thanks!

File: Makefile [hunk 9]
```diff
@@ -2,13 +2,16 @@
...
+all :: black flake8 pytest clean_docs examples html
 
 black ::
 	black .
 
 clean_docs ::
 	make -C docs clean
+	rm -rf docs/source/examples/notebooks/geo_*
+	rm -rf docs/source/examples/notebooks/tst_*
+	rm -rf docs/source/examples/notebooks/var_*
```
[the remark this answers is not in the document]
Comment (prjemian) [reply Makefile:14]: We still need it.  Each notebook can have a `NOTEBOOK_files` directory with the images from the plots generated.

File: hkl/diffract.py [hunk 10]
```diff
@@ -34,7 +34,70 @@
...
+        self.high_limit = float(high_limit)
+        self.value = float(value)
+        self.fit = bool(fit)
+
+        self._fields = "low_limit high_limit value fit".split()
+        # fmt: off
+        self._repr_fmt = (
+            "("
+            + ", ".join(
+                name + "=" + repr(getattr(self, name))
```
[the remark this answers is not in the document]
Comment (prjemian) [reply hkl/diffract.py:67]: Agreed, this looks messier than it needs to be.

File: hkl/diffract.py (hunk 1 above)
Comment (prjemian) [reply hkl/diffract.py:91]: Could prepend `_` to suggest it is not a public interface.

File: hkl/diffract.py (hunk 3 above)
Comment (prjemian) [reply hkl/diffract.py:123]: copy+paste typo

File: hkl/diffract.py (hunk 4 above)
Comment (prjemian) [reply hkl/diffract.py:129]: op cit.

File: hkl/tests/test_save_restore_UB.py (hunk 6 above)
Comment (prjemian) [reply hkl/tests/test_save_restore_UB.py:140]: Great idea!

File: hkl/tests/test_save_restore_UB.py (hunk 5 above)
Comment (prjemian) [reply hkl/tests/test_save_restore_UB.py:114]: That's right.  This can be simpler.

**Pull request merged**